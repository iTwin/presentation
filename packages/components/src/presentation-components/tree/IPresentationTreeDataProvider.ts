/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import { ActiveMatchInfo, ITreeDataProvider, TreeNodeItem } from "@itwin/components-react";
import { IModelConnection } from "@itwin/core-frontend";
import { HierarchyRequestOptions, InstanceFilterDefinition, NodeKey, NodePathElement } from "@itwin/presentation-common";
import { IPresentationDataProvider } from "../common/IPresentationDataProvider";

/**
 * Presentation tree data provider.
 * @public
 */
export interface IPresentationTreeDataProvider extends ITreeDataProvider, IPresentationDataProvider {
  /**
   * Returns a [NodeKey]($presentation-common) from given [TreeNodeItem]($components-react).
   *
   * @deprecated in 4.0. Use [[isPresentationTreeNodeItem]] and [[PresentationTreeNodeItem.key]] to get [NodeKey]($presentation-common).
   */
  getNodeKey(node: TreeNodeItem): NodeKey;

  /**
   * Returns filtered node paths.
   */
  getFilteredNodePaths(filter: string): Promise<NodePathElement[]>;

  /**
   * Creates options for nodes request.
   */
  createRequestOptions(parentKey?: NodeKey, instanceFilter?: InstanceFilterDefinition): HierarchyRequestOptions<IModelConnection, NodeKey>;
}

/**
 * Filtered presentation tree data provider.
 * @public
 */
export interface IFilteredPresentationTreeDataProvider extends IPresentationTreeDataProvider {
  /**
   * Applied filter.
   */
  filter: string;
  /**
   * Returns active match for given index.
   */
  getActiveMatch(index: number): ActiveMatchInfo | undefined;
  /**
   * Counts all filter matches.
   */
  countFilteringResults(nodePaths: ReadonlyArray<Readonly<NodePathElement>>): number;
  /**
   * Checks whether node matches applied filter or not.
   */
  nodeMatchesFilter(node: TreeNodeItem): boolean;
}
