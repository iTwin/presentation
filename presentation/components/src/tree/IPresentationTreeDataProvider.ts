/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

import { ITreeDataProvider, TreeNodeItem } from "@bentley/ui-components";
import { IPresentationDataProvider } from "../common/IPresentationDataProvider";
import { NodeKey, NodePathElement } from "@bentley/presentation-common";

/**
 * Presentation tree data provider.
 * @public
 */
export interface IPresentationTreeDataProvider extends ITreeDataProvider, IPresentationDataProvider {
  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   */
  getNodeKey(node: TreeNodeItem): NodeKey;

  /**
   * Returns filtered node paths.
   */
  getFilteredNodePaths(filter: string): Promise<NodePathElement[]>;

  /**
   * Loads the hierarchy so on-demand requests and filtering works quicker
   * @alpha
   */
  loadHierarchy?(): Promise<void>;
}
