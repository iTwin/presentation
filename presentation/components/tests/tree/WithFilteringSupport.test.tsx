/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { expect } from "chai";
import { shallow, ShallowWrapper } from "enzyme";
import * as faker from "faker";
import * as moq from "@bentley/presentation-common/tests/_helpers/Mocks";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { Tree } from "@bentley/ui-components";
import withFilteringSupport, { Props } from "../../lib/tree/WithFilteringSupport";
import FilteredPresentationTreeDataProvider from "../../lib/tree/FilteredDataProvider";
import IPresentationTreeDataProvider from "../../lib/tree/IPresentationTreeDataProvider";

// tslint:disable-next-line:variable-name naming-convention
const PresentationTree = withFilteringSupport(Tree);
interface State {
    filteredDataProvider?: FilteredPresentationTreeDataProvider;
}

const defaultState: State = {
    filteredDataProvider: undefined,
};

describe("Tree withFilteringSupport", () => {

    let testRulesetId: string;
    let tree: ShallowWrapper<Props, State, any>;
    let treeInstance: React.Component<Props, State, any>;
    const imodelMock = moq.Mock.ofType<IModelConnection>();
    const dataProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
    const filter = "filter";
    const onFilterAppliedMock = moq.Mock.ofType<(propsFilter?: string) => void>();

    beforeEach(() => {
        testRulesetId = faker.random.word();
        setupDataProvider(dataProviderMock, imodelMock.object, testRulesetId);
        tree = shallow(<PresentationTree dataProvider={dataProviderMock.object} filter={filter} />, { disableLifecycleMethods: true });
        treeInstance = tree.instance();
        expect(treeInstance.state.filteredDataProvider).to.be.undefined;
    });

    const setupDataProvider = (providerMock: moq.IMock<IPresentationTreeDataProvider>, imodel: IModelConnection, rulesetId: string) => {
        providerMock.reset();
        providerMock.setup((x) => x.connection).returns(() => imodel!);
        providerMock.setup((x) => x.rulesetId).returns(() => rulesetId!);
        providerMock.setup((x) => x.getFilteredNodePaths(moq.It.isAnyString())).returns(async () => []);
    };

    describe("componentDidUpdate", () => {

        beforeEach(() => {
            onFilterAppliedMock.reset();
            expect(treeInstance.componentDidUpdate).to.not.be.undefined;
            expect(treeInstance.state.filteredDataProvider).to.be.undefined;
        });

        it("creates new filtered data provider if filter changed", async () => {
            tree.setProps({ dataProvider: dataProviderMock.object, filter, onFilterApplied: onFilterAppliedMock.object });
            await treeInstance.componentDidUpdate!({ dataProvider: dataProviderMock.object, filter: "previous filter", onFilterApplied: onFilterAppliedMock.object }, defaultState);
            const state = treeInstance.state;
            expect(state.filteredDataProvider).to.not.be.undefined;
            expect(state.filteredDataProvider instanceof FilteredPresentationTreeDataProvider).to.be.true;
            onFilterAppliedMock.verify((x) => x(filter), moq.Times.never());
        });

        it("creates new filtered data provider if ruleset changed", async () => {
            const anotherProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
            setupDataProvider(anotherProviderMock, dataProviderMock.object.connection, faker.random.word());
            expect(anotherProviderMock.object.rulesetId).to.be.not.equal(dataProviderMock.object.rulesetId);

            tree.setProps({ dataProvider: dataProviderMock.object, filter, onFilterApplied: onFilterAppliedMock.object });
            await treeInstance.componentDidUpdate!({ dataProvider: anotherProviderMock.object, filter }, defaultState);
            const state = treeInstance.state;
            expect(state.filteredDataProvider).to.not.be.undefined;
            expect(state.filteredDataProvider instanceof FilteredPresentationTreeDataProvider).to.be.true;
            onFilterAppliedMock.verify((x) => x(filter), moq.Times.never());
        });

        it("creates new filtered data provider if connection changed", async () => {
            const anotherProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
            setupDataProvider(anotherProviderMock, moq.Mock.ofType<IModelConnection>().object, dataProviderMock.object.rulesetId);
            expect(anotherProviderMock.object.connection).to.be.not.equal(dataProviderMock.object.connection);

            tree.setProps({ dataProvider: dataProviderMock.object, filter, onFilterApplied: onFilterAppliedMock.object });
            await treeInstance.componentDidUpdate!({ dataProvider: anotherProviderMock.object, filter }, defaultState);
            const state = treeInstance.state;
            expect(state.filteredDataProvider).to.not.be.undefined;
            expect(state.filteredDataProvider instanceof FilteredPresentationTreeDataProvider).to.be.true;
            onFilterAppliedMock.verify((x) => x(filter), moq.Times.never());
        });

        it("does not create new filtered data provider if current filter is empty", async () => {
            const anotherProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
            setupDataProvider(anotherProviderMock, moq.Mock.ofType<IModelConnection>().object, faker.random.word());
            expect(anotherProviderMock.object.rulesetId).to.be.not.equal(dataProviderMock.object.rulesetId);
            expect(anotherProviderMock.object.connection).to.be.not.equal(dataProviderMock.object.connection);

            tree.setProps({ dataProvider: dataProviderMock.object, filter: "", onFilterApplied: onFilterAppliedMock.object });
            await treeInstance.componentDidUpdate!({ dataProvider: anotherProviderMock.object, filter }, defaultState);
            expect(treeInstance.state.filteredDataProvider).to.be.undefined;
            onFilterAppliedMock.verify((x) => x(""), moq.Times.once());
        });

        it("does not create new filtered data provider if current filter is undefined", async () => {
            const anotherProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
            setupDataProvider(anotherProviderMock, moq.Mock.ofType<IModelConnection>().object, faker.random.word());
            expect(anotherProviderMock.object.rulesetId).to.be.not.equal(dataProviderMock.object.rulesetId);
            expect(anotherProviderMock.object.connection).to.be.not.equal(dataProviderMock.object.connection);

            tree.setProps({ dataProvider: dataProviderMock.object, filter: undefined, onFilterApplied: onFilterAppliedMock.object });
            await treeInstance.componentDidUpdate!({ dataProvider: anotherProviderMock.object, filter }, defaultState);
            expect(treeInstance.state.filteredDataProvider).to.be.undefined;
            onFilterAppliedMock.verify((x) => x(undefined), moq.Times.once());

        });

        it("does not create new filtered data provider if filter, ruleset and connection did not change and filteredDataProvider is not set", async () => {
            tree.setProps({ dataProvider: dataProviderMock.object, filter, onFilterApplied: onFilterAppliedMock.object });
            await treeInstance.componentDidUpdate!({ dataProvider: dataProviderMock.object, filter }, defaultState);
            const filteredDataProvider = treeInstance.state.filteredDataProvider;
            expect(filteredDataProvider).to.be.undefined;
            onFilterAppliedMock.verify((x) => x(filter), moq.Times.never());
        });

        it("does not create new filtered data provider if filter, ruleset and connection did not change", async () => {
            const filteredDataProviderMock = moq.Mock.ofType<FilteredPresentationTreeDataProvider>();
            filteredDataProviderMock.setup((x) => x.rulesetId).returns(() => dataProviderMock.object.rulesetId);
            filteredDataProviderMock.setup((x) => x.connection).returns(() => dataProviderMock.object.connection);
            filteredDataProviderMock.setup((x) => x.filter).returns(() => filter);
            tree.setState({ filteredDataProvider: filteredDataProviderMock.object });
            tree.setProps({ dataProvider: dataProviderMock.object, filter, onFilterApplied: onFilterAppliedMock.object });

            await treeInstance.componentDidUpdate!({ dataProvider: dataProviderMock.object, filter }, defaultState);
            const filteredDataProvider = treeInstance.state.filteredDataProvider;
            expect(filteredDataProvider).to.be.equal(filteredDataProviderMock.object);
            onFilterAppliedMock.verify((x) => x(filter), moq.Times.once());

        });

    });

    describe("componentDidMount", () => {

        beforeEach(() => {
            onFilterAppliedMock.reset();
            expect(treeInstance.componentDidMount).to.not.be.undefined;
            expect(treeInstance.state.filteredDataProvider).to.be.undefined;
        });

        it("creates new filtered data provider if filter is not empty or undefined", async () => {
            tree.setProps({ dataProvider: dataProviderMock.object, filter, onFilterApplied: onFilterAppliedMock.object });
            await treeInstance.componentDidMount!();

            const state = treeInstance.state;
            expect(state.filteredDataProvider).to.not.be.undefined;
            expect(state.filteredDataProvider instanceof FilteredPresentationTreeDataProvider).to.be.true;
            onFilterAppliedMock.verify((x) => x(filter), moq.Times.never());
        });

        it("does not create new filtered data provider if filter is empty", async () => {
            tree.setProps({ dataProvider: dataProviderMock.object, filter: "", onFilterApplied: onFilterAppliedMock.object });
            await treeInstance.componentDidMount!();

            const state = treeInstance.state;
            expect(state.filteredDataProvider).to.be.undefined;
            onFilterAppliedMock.verify((x) => x(""), moq.Times.once());
        });

        it("does not create new filtered data provider if filter is undefined", async () => {
            tree.setProps({ dataProvider: dataProviderMock.object, filter: undefined, onFilterApplied: onFilterAppliedMock.object });
            await treeInstance.componentDidMount!();

            const state = treeInstance.state;
            expect(state.filteredDataProvider).to.be.undefined;
            onFilterAppliedMock.verify((x) => x(undefined), moq.Times.once());
        });

        it("does not call onFilterApplied if it is not set", async () => {
            tree.setProps({ dataProvider: dataProviderMock.object, filter: undefined });
            await treeInstance.componentDidMount!();

            onFilterAppliedMock.verify((x) => x(filter), moq.Times.never());
        });

    });

    describe("getDerivedStateFromProps", () => {
        const getDerivedStateFromProps: (nextProps: Props, state: State) => State = (PresentationTree as any).getDerivedStateFromProps;
        const filteredDataProviderMock = moq.Mock.ofType<FilteredPresentationTreeDataProvider>();

        before(() => {
            expect(getDerivedStateFromProps).to.not.be.undefined;
        });

        it("does not change filteredDataProvider if filter is not empty or undefined", async () => {
            const state = { filteredDataProvider: filteredDataProviderMock.object, displayOverlay: false };

            const result = getDerivedStateFromProps({ dataProvider: dataProviderMock.object, filter: "filter" }, state);
            expect(result.filteredDataProvider).to.be.equal(state.filteredDataProvider);
        });

        it("clears filteredDataProvider if filter is empty", async () => {
            const state = { filteredDataProvider: filteredDataProviderMock.object, displayOverlay: false };

            const result = getDerivedStateFromProps({ dataProvider: dataProviderMock.object, filter: "" }, state);
            expect(result.filteredDataProvider).to.be.undefined;
        });

        it("clears filteredDataProvider if filter is undefined", async () => {
            const state = { filteredDataProvider: filteredDataProviderMock.object, displayOverlay: false };

            const result = getDerivedStateFromProps({ dataProvider: dataProviderMock.object, filter: undefined }, state);
            expect(result.filteredDataProvider).to.be.undefined;
        });

    });

    describe("loadDataProvider", () => {
        it("does not set state if filter changed while loading provider", async () => {
            const anotherDataProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
            anotherDataProviderMock.setup((x) => x.getFilteredNodePaths(moq.It.isAnyString()))
                .callback(() => { tree.setProps({ dataProvider: anotherDataProviderMock.object, filter: "different filter" }); })
                .returns(async () => []);
            tree.setProps({ dataProvider: anotherDataProviderMock.object, filter });
            expect(treeInstance.componentDidMount).to.not.be.undefined;
            expect(treeInstance.state.filteredDataProvider).to.be.undefined;

            await treeInstance.componentDidMount!();
            expect(treeInstance.state.filteredDataProvider).to.be.undefined;
        });

        it("calls onHighlightedCounted if it's given through props", async () => {
            let functionCalled = false;
            tree.setProps({ onHighlightedCounted: () => { functionCalled = true; } });

            if (treeInstance.componentDidMount)
                await treeInstance.componentDidMount();

            expect(functionCalled).to.be.true;
        });
    });

    describe("render", () => {

        it("passes expanded nodes", () => {
            const filteredDataProviderMock = moq.Mock.ofType<FilteredPresentationTreeDataProvider>();
            filteredDataProviderMock.setup((x) => x.getAllNodeIds()).returns(() => [faker.random.word(), faker.random.word()]);
            tree.setState({ filteredDataProvider: filteredDataProviderMock.object });
            expect(treeInstance.render()).to.matchSnapshot();
        });

        it("renders without overlay if filter is undefined", () => {
            tree.setState({ filteredDataProvider: undefined });
            tree.setProps({ filter: undefined, dataProvider: dataProviderMock.object });
            expect(treeInstance.render()).to.matchSnapshot();
        });

        it("renders without overlay if filter is empty", () => {
            tree.setState({ filteredDataProvider: undefined });
            tree.setProps({ filter: "", dataProvider: dataProviderMock.object });
            expect(treeInstance.render()).to.matchSnapshot();
        });

        it("renders with overlay if filter is different", () => {
            const filteredDataProviderMock = moq.Mock.ofType<FilteredPresentationTreeDataProvider>();
            filteredDataProviderMock.setup((x) => x.rulesetId).returns(() => dataProviderMock.object.rulesetId);
            filteredDataProviderMock.setup((x) => x.connection).returns(() => dataProviderMock.object.connection);
            filteredDataProviderMock.setup((x) => x.filter).returns(() => "different filter");

            tree.setState({ filteredDataProvider: filteredDataProviderMock.object });
            tree.setProps({ filter, dataProvider: dataProviderMock.object });

            expect(treeInstance.render()).to.matchSnapshot();
        });

        it("renders with overlay if rulesetId is different", async () => {
            const differentRulesetId = faker.random.word();
            expect(differentRulesetId).to.not.be.equal(dataProviderMock.object.rulesetId);
            const filteredDataProviderMock = moq.Mock.ofType<FilteredPresentationTreeDataProvider>();
            filteredDataProviderMock.setup((x) => x.rulesetId).returns(() => differentRulesetId);
            filteredDataProviderMock.setup((x) => x.connection).returns(() => dataProviderMock.object.connection);
            filteredDataProviderMock.setup((x) => x.filter).returns(() => filter);

            tree.setState({ filteredDataProvider: filteredDataProviderMock.object });
            tree.setProps({ filter, dataProvider: dataProviderMock.object });

            expect(treeInstance.render()).to.matchSnapshot();
        });

        it("renders with overlay if connection is different", async () => {
            const anotherConnectionMock = moq.Mock.ofType<IModelConnection>();
            const filteredDataProviderMock = moq.Mock.ofType<FilteredPresentationTreeDataProvider>();
            filteredDataProviderMock.setup((x) => x.rulesetId).returns(() => dataProviderMock.object.rulesetId);
            filteredDataProviderMock.setup((x) => x.connection).returns(() => anotherConnectionMock.object);
            filteredDataProviderMock.setup((x) => x.filter).returns(() => filter);

            tree.setState({ filteredDataProvider: filteredDataProviderMock.object });
            tree.setProps({ filter, dataProvider: dataProviderMock.object });

            expect(treeInstance.render()).to.matchSnapshot();
        });

        it("renders without overlay if nothing is different", async () => {
            const filteredDataProviderMock = moq.Mock.ofType<FilteredPresentationTreeDataProvider>();
            filteredDataProviderMock.setup((x) => x.rulesetId).returns(() => dataProviderMock.object.rulesetId);
            filteredDataProviderMock.setup((x) => x.connection).returns(() => dataProviderMock.object.connection);
            filteredDataProviderMock.setup((x) => x.filter).returns(() => filter);

            tree.setState({ filteredDataProvider: filteredDataProviderMock.object });
            tree.setProps({ filter, dataProvider: dataProviderMock.object });

            expect(treeInstance.render()).to.matchSnapshot();
        });

        it("renders with full highlightingProps", async () => {
            const filteredDataProviderMock = moq.Mock.ofType<FilteredPresentationTreeDataProvider>();
            filteredDataProviderMock.setup((x) => x.getActiveResultNode(moq.It.isAny())).returns(() => ({ id: "test", index: 0 }));

            tree.setState({ filteredDataProvider: filteredDataProviderMock.object });
            tree.setProps({ activeHighlightedIndex: 6 });

            expect(treeInstance.render()).to.matchSnapshot();
        });
    });

});
