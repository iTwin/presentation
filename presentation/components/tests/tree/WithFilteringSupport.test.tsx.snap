// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree withFilteringSupport render passes expanded nodes 1`] = `
<div
  className="filteredTree"
>
  <DataTree
    dataProvider={null}
    expandedNodes={
      Array [
        "Solomon Islands Dollar",
        "Focused",
      ]
    }
    nodeHighlightingProps={
      Object {
        "activeResultNode": undefined,
        "searchText": "filter",
      }
    }
  />
  <div
    className="filteredTreeOverlay"
  />
</div>
`;

exports[`Tree withFilteringSupport render renders with full highlightingProps 1`] = `
<div
  className="filteredTree"
>
  <DataTree
    activeHighlightedIndex={6}
    dataProvider={null}
    expandedNodes={undefined}
    nodeHighlightingProps={
      Object {
        "activeResultNode": Object {
          "id": "test",
          "index": 0,
        },
        "searchText": "filter",
      }
    }
  />
  <div
    className="filteredTreeOverlay"
  />
</div>
`;

exports[`Tree withFilteringSupport render renders with overlay if connection is different 1`] = `
<div
  className="filteredTree"
>
  <DataTree
    dataProvider={null}
    expandedNodes={undefined}
    nodeHighlightingProps={
      Object {
        "activeResultNode": undefined,
        "searchText": "filter",
      }
    }
  />
  <div
    className="filteredTreeOverlay"
  />
</div>
`;

exports[`Tree withFilteringSupport render renders with overlay if filter is different 1`] = `
<div
  className="filteredTree"
>
  <DataTree
    dataProvider={null}
    expandedNodes={undefined}
    nodeHighlightingProps={
      Object {
        "activeResultNode": undefined,
        "searchText": "filter",
      }
    }
  />
  <div
    className="filteredTreeOverlay"
  />
</div>
`;

exports[`Tree withFilteringSupport render renders with overlay if rulesetId is different 1`] = `
<div
  className="filteredTree"
>
  <DataTree
    dataProvider={null}
    expandedNodes={undefined}
    nodeHighlightingProps={
      Object {
        "activeResultNode": undefined,
        "searchText": "filter",
      }
    }
  />
  <div
    className="filteredTreeOverlay"
  />
</div>
`;

exports[`Tree withFilteringSupport render renders without overlay if filter is empty 1`] = `
<div
  className="filteredTree"
>
  <DataTree
    dataProvider={null}
    expandedNodes={Array []}
    nodeHighlightingProps={
      Object {
        "activeResultNode": undefined,
        "searchText": "",
      }
    }
  />
</div>
`;

exports[`Tree withFilteringSupport render renders without overlay if filter is undefined 1`] = `
<div
  className="filteredTree"
>
  <DataTree
    dataProvider={null}
    expandedNodes={Array []}
    nodeHighlightingProps={
      Object {
        "activeResultNode": undefined,
        "searchText": undefined,
      }
    }
  />
</div>
`;

exports[`Tree withFilteringSupport render renders without overlay if nothing is different 1`] = `
<div
  className="filteredTree"
>
  <DataTree
    dataProvider={null}
    expandedNodes={undefined}
    nodeHighlightingProps={
      Object {
        "activeResultNode": undefined,
        "searchText": "filter",
      }
    }
  />
</div>
`;
